src/webgldrawer.js

```diff
@@ -0,0 +1,1235 @@
+                let matrixArray = new Array(maxTextures);
+                let opacityArray = new Array(maxTextures);
+
+                // iterate over tiles and add data for each one to the buffers
+                for(let tileIndex = 0; tileIndex < tilesToDraw.length; tileIndex++){
+                    let tile = tilesToDraw[tileIndex].tile;
+                    let index = tileIndex % maxTextures;
+                    let tileContext = tile.getCanvasContext();
+
+                    let textureInfo = tileContext ? this._TextureMap.get(tileContext.canvas) : null;
+                    if(textureInfo){
+                        this._getTileData(tile, tiledImage, textureInfo, overallMatrix, index, texturePositionArray, textureDataArray, matrixArray, opacityArray);
+                    } else {
+                        // console.log('No tile info', tile);
+                    }
+                    if( (index === maxTextures - 1) || (tileIndex === tilesToDraw.length - 1)){
+                        // We've filled up the buffers: time to draw this set of tiles
+
+                        // bind each tile's texture to the appropriate gl.TEXTURE#
+                        for(let i = 0; i <= index; i++){
+                            gl.activeTexture(gl.TEXTURE0 + i);
+                            gl.bindTexture(gl.TEXTURE_2D, textureDataArray[i]);
+                        }
+
+                        // set the buffer data for the texture coordinates to use for each tile
+                        gl.bindBuffer(gl.ARRAY_BUFFER, this._firstPass.bufferTexturePosition);
+                        gl.bufferData(gl.ARRAY_BUFFER, texturePositionArray, gl.DYNAMIC_DRAW);
+
+                        // set the transform matrix uniform for each tile
+                        matrixArray.forEach( (matrix, index) => {
+                            gl.uniformMatrix3fv(this._firstPass.uTransformMatrices[index], false, matrix);
+                        });
+                        // set the opacity uniform for each tile
+                        gl.uniform1fv(this._firstPass.uOpacities, new Float32Array(opacityArray));
+
+                        // bind vertex buffers and (re)set attributes before calling gl.drawArrays()
+                        gl.bindBuffer(gl.ARRAY_BUFFER, this._firstPass.bufferOutputPosition);
+                        gl.vertexAttribPointer(this._firstPass.aOutputPosition, 2, gl.FLOAT, false, 0, 0);
+
+                        gl.bindBuffer(gl.ARRAY_BUFFER, this._firstPass.bufferTexturePosition);
+                        gl.vertexAttribPointer(this._firstPass.aTexturePosition, 2, gl.FLOAT, false, 0, 0);
+
+                        gl.bindBuffer(gl.ARRAY_BUFFER, this._firstPass.bufferIndex);
+                        gl.vertexAttribPointer(this._firstPass.aIndex, 1, gl.FLOAT, false, 0, 0);
+
+                        // Draw! 6 vertices per tile (2 triangles per rectangle)
+                        gl.drawArrays(gl.TRIANGLES, 0, 6 * (index + 1) );
```

Comment:
> I don't think these 'batches' save much

I've seen lots of places that reducing the number of draw calls is critical for webgl performance (and experimenting with the performance testing demo bears this out). Batching even 16 tiles into a single draw reduces overall draws by a factor of 10 or more...

Reply:
> via newer version or extensions

I agree this would likely improve efficiency further. As in other places of the discussion, the question is what do we want to be the minimal webgl support we can work with.

Reply:
> multidraw really does as it reduces the number of API calls

I'm not familiar with performance improvements of `multidraw` but my understanding is that `draw` calls and sending texture data to the GPU is expensive, but many API calls like binding buffers and such are relatively inexpensive.

Reply:
> `draw` calls and sending texture data to the GPU is expensive
Ah, I misunderstood the for loop. It is a bit misleading since you use `index` var to specify the number of vertices to draw which is confusing. I don't know how much binding expensive is, maybe it would be relevant only if we avoided all binding calls (since CPU - GPU sync is certainly expensive)...

Reply:
I renamed some variables to make this more clear.